### ISSUE TITLE
Animating map view contraints is jerky and does not update properly

### ISSUE THREAD

**Issue by friedbunny** [mapbox/mapbox-gl-native#4352]
The proper way to animate MGLMapView is to modify its auto layout constraints (and not the frame), but this doesn’t behave correctly:
- White gaps at top and bottom jump as the map frame instantly shrinks to fit, rather than animate to the new, smaller constraint. The logo and attribution subviews do animate correctly, which indicates this is an issue with the GLKView that provides the map itself.
- User location annotation does not move, it just sits around waiting to be in the correct spot again.

![animating-constraint-jumping](https://cloud.githubusercontent.com/assets/1198851/13827612/fc0ed78e-eb93-11e5-9fba-411f4b7c0cae.gif)

Preview of the [test project](https://github.com/mapbox/mapbox-gl-native/files/176840/animationtest.zip):

![screen shot 2016-03-16 at 3 10 26 pm](https://cloud.githubusercontent.com/assets/1198851/13827615/feb32922-eb93-11e5-8e25-b4e7831fa6bd.png)

``` swift
import UIKit
import Mapbox

class ViewController: UIViewController {

    @IBOutlet weak var mapView: MGLMapView!
    @IBOutlet weak var heightConstraint: NSLayoutConstraint!

    override func viewDidLoad() {
        super.viewDidLoad()
        mapView.showsUserLocation = true
    }

    @IBAction func animate(sender: AnyObject) {
        UIView.animateWithDuration(1.5, delay: 0, options: .CurveEaseOut, animations: {
            self.mapView.userInteractionEnabled = false

            self.heightConstraint.constant = self.heightConstraint.constant > 200 ? 200 : self.view.frame.height/2

            self.view.layoutIfNeeded()
        }, completion: { finished in
            self.mapView.userInteractionEnabled = true
        })
    }

}
```

/cc @1ec5 @boundsj

**Comment by harri121** [comment]
+1

**Comment by harri121** [comment]
Is there a workaround?

**Comment by 1ec5** [comment]
I’m not sure about a workaround, but I suspect the problem may be that `-layoutSubviews` only gets called at the beginning of the animation, so `mbgl::Map::update()` gets called only once with the final size. Meanwhile, the user dot appears to stay stationary because it’s actually animating in the opposite direction to compensate. (If the map view is flush with the bottom of the view controller, showing and hiding the bottom bar will cause the user dot to visibly slide into place.)

**Comment by 1ec5** [comment]
The fix is probably to move the call to `mbgl::Map::update()` from `-layoutSubviews` to the `-setFrame:` or `-setBounds:` override and also have `MBGLView::getSize()` to return GLKView’s presentation layer’s size instead of MGLMapView’s explicit size.

**Comment by harri121** [comment]
Ok, do you know when this will be fixed?

**Comment by afinlayson** [comment]
Looking for this as well. 
Thanks

**Comment by lilykaiser** [comment]
Closing due to ticket age. If this is still a problem, please re-open

**Comment by katiesmillie** [comment]
@friedbunny I'm also having this issue and would appreciate an update. Are there any workarounds?

**Comment by 1ec5** [comment]
As a possible starting point, we should check whether `-[GLKView drawableHeight]` is even animatable:

https://github.com/mapbox/mapbox-gl-native/blob/d5f0e4b41e0ece7795a350cc6b9b68fc3aa356ce/platform/ios/src/MGLMapView.mm#L615-L617.

**Comment by katiesmillie** [comment]
This is still an issue that I've been following and hoping for a resolution. I hope you'll consider re-opening it. 

The best workaround that I've come up with is to leave the size of the mapview unchanged, and change the content insets of the map (which adjusts the coordinates bounds and center of the map). The map animation happens after the second view has been finished animating so you lose the effect you see above where they both animate together. This significantly degrades the experience as it has become a very common mobile UI pattern with all the major map apps.

**Comment by julianrex** [comment]
This is still open @katiesmillie! 

Your workaround of changing the content insets may actually be part of the solution that we're looking at (or similar) - since adjusting the frame buffers during an animation gives poor performance.

**Comment by racer1988** [comment]
+1 for this issue

**Comment by racer1988** [comment]
The same issue is also happening when the iOS tab bar or Navigation Bar are showing animated. as the container of the map will change constrains animated.

I also noticed that changing the content inset (animated if I remember correctly) also does not redraw the current location annotation position (as originally reported)

**Comment by 1ec5** [comment]
The root cause of this issue is that `MGLMapView.camera`, `centerCoordinate`, etc. are not implicitly animatable. Supporting implicit animation for the camera properties would be one side effect of reimplementing camera animation atop Core Animation: #9808.

**Comment by 1ec5** [comment]
Replacing GLKView with a built-in Core Animation layer (#14785) would make the effect somewhat less jarring, as it is on macOS.

**Comment by racer1988** [comment]
@1ec5 Would it be possible and is the idea to support animation for the camera and centerCoordinate etc?

We noticed that changing the inset/frame of the map while zooming the camera will:

trigger `uiView SafeAreaInsetChanged`

`mapView.adjustContentInset`

which calls `MGLMapView.cancelTransitions` that also destroy the camera anmations

**Comment by 1ec5** [comment]
> Would it be possible and is the idea to support animation for the camera and centerCoordinate etc?

Yes, I think implicit animation of these properties is essentially what this issue tracks. There is a separate effort to overhaul the camera APIs on both Android and iOS/macOS, which is sort of tracked in #9808, but I think we have to solve this issue regardless, especially considering the prominence of implicit animation in SwiftUI.

**Comment by robert-frankin** [comment]
Hi, I am getting same issue in iOS. Is there any solution to fix it? Please advise.
